File: functions/lib/functions/src/test/java/com/functions/events/utils/EventIdGeneratorTest.java
package com.functions.events.utils;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotEquals;
import static org.junit.Assert.assertTrue;

import java.time.Instant;

import org.junit.Test;

import com.github.f4b6a3.ulid.Ulid;

public class EventIdGeneratorTest {
    private static final String EVENT_ID_PREFIX = "evt_";

    @Test
    public void createsPrefixedUlidEventIds() {
        String eventId = EventIdGenerator.newEventId();

        assertTrue(eventId.startsWith(EVENT_ID_PREFIX));
        Ulid ulid = Ulid.from(eventId.substring(EVENT_ID_PREFIX.length()));
        assertEquals(26, ulid.toString().length());
        assertTrue(Math.abs(ulid.getInstant().toEpochMilli() - Instant.now().toEpochMilli()) < 5_000);
    }

    @Test
    public void createsDistinctOrderedEventIds() {
        String first = EventIdGenerator.newEventId();
        String second = EventIdGenerator.newEventId();

        assertNotEquals(first, second);
        assertTrue(first.compareTo(second) < 0);
    }
}

File: frontend/services/tests/eventIdGenerator.test.ts
import { decodeTime } from "ulid";
import { newEventId } from "../src/events/eventsUtils/eventIdGenerator";

const EVENT_ID_PATTERN = /^evt_[0-9A-HJKMNP-TV-Z]{26}$/;

describe("newEventId", () => {
  it("returns evt_ plus a ULID", () => {
    const eventId = newEventId();

    expect(eventId).toMatch(EVENT_ID_PATTERN);
    expect(Math.abs(decodeTime(eventId.slice("evt_".length)) - Date.now())).toBeLessThan(5_000);
  });

  it("returns distinct, time-ordered ids", () => {
    const first = newEventId();
    const second = newEventId();

    expect(second).not.toBe(first);
    expect(first < second).toBe(true);
  });
});
